torch/csrc/dynamo/eval_frame.c

```diff
@@ -120,6 +120,31 @@ static PyObject* THPPyInterpreterFrame_f_back(THPPyInterpreterFrame* self, PyObj
 #endif // IS_PYTHON_3_11_PLUS
 }
 
+static PyObject* THPPyInterpreterFrame_closure(THPPyInterpreterFrame* self, PyObject* _noargs) {
+#if IS_PYTHON_3_12_PLUS
+  PyObject* closure = ((PyFunctionObject*) self->frame->f_funcobj)->func_closure;
+  return closure == NULL ? PyTuple_New(0) : Py_XNewRef(closure);
+#elif IS_PYTHON_3_11_PLUS
+  PyObject* closure = ((PyFunctionObject*) self->frame->f_func)->func_closure;
+  return closure == NULL ? PyTuple_New(0) : Py_XNewRef(closure);
+#else
+  PyCodeObject* code = self->frame->f_code;
+  // Why this check? See
+  // https://github.com/python/cpython/blob/5f24da9d75bb0150781b17ee4706e93e6bb364ea/Objects/frameobject.c#L1058-L1065
+  if (code->co_flags & CO_OPTIMIZED) {
```

Comment:
There's some code duplication with https://github.com/pytorch/pytorch/blob/27c7caf745e103fd1952791dac6d59720e1df615/torch/csrc/dynamo/framelocals_mapping.cpp#L114, but it's not a huge deal